unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -515,4 +522,45 @@ public int hashCode() {
     }
     return result;
   }
+
+  /**
+   * Encodes a string into a Soundex value. Soundex is an encoding used to relate similar names,
+   * but can also be used as a general purpose scheme to find word with similar phonemes.
+   */
+  public UTF8String Soundex() {
+    if (numBytes == 0) {
+      return UTF8String.fromBytes(new byte[0]);
+    }
+    String tmp;
+    byte data[] = {'0','0','0','0'};
+    char ch;
+    int idx = 0;
+    int offset = numBytesForFirstByte(getByte(0));
+    if(offset>1)
+    {
+        return UTF8String.fromBytes(getBytes());
+    }
+    int i = 1;
+    int j = 1;
+    data[0] = getByte(0);
+
+    while (i < numBytes) {
+      if (j > 4) break;
+      idx = getByte(i) - 'A';
+      if (idx >= 0 && idx <= US_ENGLISH_MAPPING.length) {
+        ch = US_ENGLISH_MAPPING[idx];
+        if (ch - '0' > 0) {
+          tmp = Character.toString(ch);
+          System.arraycopy(tmp.getBytes(), 0, data, j, 1);
```

Comment:
Instead of converting here `char` to `byte` change the type of `US_ENGLISH_MAPPING` to `private static final byte[] US_ENGLISH_MAPPING`.